src/na/device/Generator.cpp

```diff
@@ -0,0 +1,380 @@
+
+    // approximate indices of the top right corner
+    const auto& [topRightI, topRightJ] = solve2DLinearEquation<int64_t>(
+        baseVector1X, baseVector2X, baseVector1Y, baseVector2Y,
+        extentOriginX + extentWidth - latticeOriginX,
+        extentOriginY + extentHeight - latticeOriginY);
+
+    const auto minI = static_cast<int64_t>(
+        std::floor(std::min({bottomLeftI, bottomRightI, topLeftI, topRightI})));
+    const auto minJ = static_cast<int64_t>(
+        std::floor(std::min({bottomLeftJ, bottomRightJ, topLeftJ, topRightJ})));
+    const auto maxI = static_cast<int64_t>(
+        std::floor(std::max({bottomLeftI, bottomRightI, topLeftI, topRightI})));
+    const auto maxJ = static_cast<int64_t>(
+        std::floor(std::max({bottomLeftJ, bottomRightJ, topLeftJ, topRightJ})));
+
+    const std::vector limits{maxI, maxJ};
+    std::vector indices{minI, minJ};
+    for (bool loop = true; loop;
+         loop = increment(indices, limits), ++subModuleCount) {
+      // For every sublattice offset, add a site for repetition indices
+      for (const auto& offset : lattice.sublatticeOffsets) {
+        const auto id = count++;
+        auto x = latticeOriginX + offset.x;
+        auto y = latticeOriginY + offset.y;
+        x += indices[0] * baseVector1X;
+        y += indices[0] * baseVector1Y;
+        x += indices[1] * baseVector2X;
+        y += indices[1] * baseVector2Y;
+        if (extentOriginX <= x && x < extentOriginX + extentWidth &&
+            extentOriginY <= y && y < extentOriginY + extentHeight) {
+          // Only add the site if it is within the extent of the lattice
+          os << "\\\n  "
+                "var.emplace_back(std::make_unique<MQT_NA_QDMI_Site_impl_d>("
+                "MQT_NA_QDMI_Site_impl_d{"
+             << id << ", " << moduleCount << ", " << subModuleCount << ", "
+             << static_cast<double>(x) * lengthUnit << ", "
+             << static_cast<double>(y) * lengthUnit << "}));";
+        }
+      }
+    }
+    ++moduleCount;
+  }
+  os << "\n";
+}
+
+/**
+ * @brief Writes the operations from the device object.
+ * @param device is the device object containing the operations configuration.
+ * @param timeUnit is the time unit to use for the operations durations.
+ * @param os is the output stream to write the operations to.
+ */
+auto writeOperations(const Device& device, const double timeUnit,
+                     std::ostream& os) -> void {
+  os << "#define INITIALIZE_OPERATIONS(var) var.clear();";
+  for (const auto& operation : device.globalSingleQubitOperations) {
+    os << "\\\n"
+          "  var.emplace_back(std::make_unique<MQT_NA_QDMI_Operation_impl_d>("
+          "MQT_NA_QDMI_Operation_impl_d{\""
+       << operation.name
+       << "\", MQT_NA_QDMI_Operation_impl_d::Type::GlobalSingleQubit, "
+       << operation.numParameters << ", 1, "
+       << static_cast<double>(operation.duration) * timeUnit << ", "
+       << operation.fidelity << "}));";
```

Comment:
I think the patterns throughout this file can be simplified a little bit because `std::make_unique` forwards arguments. Something like the following should work
```suggestion
    os << "\\\n"
          "  var.emplace_back(std::make_unique<MQT_NA_QDMI_Operation_impl_d>(\""
       << operation.name
       << "\", MQT_NA_QDMI_Operation_impl_d::Type::GlobalSingleQubit, "
       << operation.numParameters << ", 1, "
       << static_cast<double>(operation.duration) * timeUnit << ", "
       << operation.fidelity << "));";
```
It might be that even the `std::make_unique` is superfluous